File: c/src/transitionfunction.c
#include "transitionfunction.h"
#include <math.h>
#include <stdio.h>

void tfunc_state_init(struct tfunc_state_s *tfs)
{
	tfs->alpham = 0.147;
	tfs->alphan = 0.028;
	
	tfs->bint[0] = 0.278;
	tfs->bint[1] = 0.365;
	
	tfs->dint[0] = 0.267;
	tfs->dint[1] = 0.445;
}

float tfunc_sig1(const struct tfunc_state_s *tfs, float x, float a, float alpha)
{
    float sig1_val = 1 / (1 + expf(-1 * (x - a) / (4*alpha)));
    //printf("-D- sig1: sig1_val = %f\n", sig1_val);
	return sig1_val;
}

float tfunc_sig2(const struct tfunc_state_s *tfs, float x, float a, float b, float alpha)
{
    float sig1a = tfunc_sig1(tfs, x, a, alpha);
    float sig1b = tfunc_sig1(tfs, x, b, alpha);
    float sig2_val = sig1a * (1 - sig1b);
    //printf("-D- sig2: sig1a = %f, sig1b = %f, sig2_val = %f\n", sig1a, sig1b, sig2_val);

	return sig1a * (1 - sig1b);
}

float tfunc_sig3(const struct tfunc_state_s *tfs, float x, float y, float m, float alpha)
{
	float sig1m = tfunc_sig1(tfs, m, 0.5, alpha);
	float sig3_val = x * (1 - sig1m) + y * sig1m;
    //printf("-D- sig3: sig3_val = %f, x = %f, y = %f\n", sig3_val, x, y);

	return sig3_val;
}

float tfunc_apply(const struct tfunc_state_s *tfs, float n, float m)
{
	float b0 = tfs->bint[0];
	float b1 = tfs->bint[1];
	float d0 = tfs->dint[0];
	float d1 = tfs->dint[1];

	float a = tfunc_sig3(tfs, b0, d0, m, tfs->alpham);
	float b = tfunc_sig3(tfs, b1, d1, m, tfs->alpham);
    float sig2_val = tfunc_sig2(tfs, n, a, b, tfs->alphan);

    //printf("-D- tfunc_apply: sig2_val = %f, a = %f, b = %f, m = %f, n = %f\n", sig2_val, a, b, m, n);

	return sig2_val;
}

void tfunc_set_alpham(struct tfunc_state_s *tfs, float alpham)
{
	tfs->alpham = alpham;
}

void tfunc_set_alphan(struct tfunc_state_s *tfs, float alphan)
{
	tfs->alphan = alphan;
}

void tfunc_set_birth_interval(struct tfunc_state_s *tfs, float b0, float b1)
{
	tfs->bint[0] = b0;
	tfs->bint[1] = b1;
}

void tfunc_set_death_interval(struct tfunc_state_s *tfs, float d0, float d1)
{
	tfs->dint[0] = d0;
	tfs->dint[1] = d1;
}

File: python/transitionfunction.py
import numpy as np
import matplotlib
import matplotlib.pyplot as plt


class TransitionFunction:

    def __init__(self, alpham=0.147, alphan=0.028, bint=[0.278, 0.365], dint=[0.267, 0.445], dtype=np.float64):
        self.alpham = dtype(alpham)
        self.alphan = dtype(alphan)
        self.bint = dtype(bint)
        self.dint = dtype(dint)

    def sig1(self, x, a, alpha):
        return 1 / (1 + np.exp(-4*(x - a) / alpha))

    def sig2(self, x, a, b, alpha):
        return self.sig1(x, a, alpha) * (1 - self.sig1(x, b, alpha))

    def sig3(self, x, y, m, alpha):
        sig1m = self.sig1(m, 0.5, alpha)

        return x * (1 - sig1m) + y * sig1m

    def __call__(self, n, m):
        b1, b2 = self.bint
        d1, d2 = self.dint

        a = self.sig3(b1, d1, m, self.alpham)
        b = self.sig3(b2, d2, m, self.alpham)

        return self.sig2(n, a, b, self.alphan)

    def plot(self, npts=100, show=True, save=False, savefile='transition_function.png'):
        p0, p1 = dtype([0, 1])
        pts = np.linspace(p0, p1, npts)
        XX, YY = np.meshgrid(pts, pts)
        ZZ = self.__call__(XX, YY)

        fig, ax = plt.subplots(1, subplot_kw={'projection': '3d'})
        surf = ax.plot_surface(XX, YY, ZZ, cmap=matplotlib.cm.coolwarm, linewidth=0.5, antialiased=False)

        # Customize the z axis.
        ax.set_xlabel('Annular Average')
        ax.set_ylabel('Circular Average')

        # Add a color bar which maps values to colors.
        fig.colorbar(surf, shrink=0.5, aspect=5)

        if show:
            plt.show()

        if save:
            fig.savefig(savefile)
